synapse/rest/client/v2_alpha/report_event.py

```diff
@@ -31,16 +44,39 @@
 class ReportEventRestServlet(RestServlet):
     PATTERNS = client_patterns("/rooms/(?P<room_id>[^/]*)/report/(?P<event_id>[^/]*)$")
 
+    # As we cannot import HomeServer without creating a circular dependency,
+    # use a forward reference.
+    hs: 'HomeServer'
+    auth: Auth
+    federation_sender: Optional[FederationSender]
+    server_notices: ServerNoticesManager
+    state: StateHandler
+    storage: Storage
+    store: DataStore
+
     def __init__(self, hs):
         super().__init__()
         self.hs = hs
         self.auth = hs.get_auth()
+
+        self.federation_sender = None
+        if hs.should_send_federation():
+            self.federation_sender = hs.get_federation_sender()
+            hs.get_federation_registry().register_edu_handler(
+                "org.matrix.m.content_report", self.on_receive_report_through_federation
+            )
         self.clock = hs.get_clock()
+        self.federation_client = hs.get_federation_client()
+        self.server_notices = hs.get_server_notices_manager()
+        self.state = hs.get_state_handler()
         self.store = hs.get_datastore()
+        self.storage = hs.get_storage()
```

Comment:
the line between "what goes in `synapse.rest.client`" and "what goes in `synapse.handlers`" is fuzzy, but if you're dealing with federation stuff, that's a pretty clear sign to me that stuff needs factoring out into `synapse.handlers`.

I suggest making an `AbuseReportHandler`, which can be called from the on_POST method here, and can contain the EDU handler.